platform/engineering/design-docs/2020-08-17-mobile-api-design-doc.md

```diff
@@ -0,0 +1,173 @@
+
+
+### Code Location
+The code will live in a 'mobile' directory within the [modules](https://github.com/department-of-veterans-affairs/vets-api/tree/master/modules) directory within vets-api.
+
+### Testing Plan
+To ensure changes to shared backend services will not break code, tests will be written to assert the mobile API requirements. Any breaking changes to upstream services will fail our tests.
+
+### Logging
+We plan to adhere to all VSP standards for logging.
+
+### Debugging
+TBD
+
+### Caveats
+TBD
+
+### Security Concerns
+#### Potential threats
+
+##### SSOe introspect DoS
+Adding more surface area to the API does enable new possibilities for denial-of-service attacks. We plan on making a call to SSOe (introspect) with every vets-api session cache miss, to validate the access token and establish a session. An attacker could DoS the introspect endpoint by passing multiple requests all with invalid access tokens.
+
+Remediation could include caching each “miss” for a period of time, protecting the introspect endpoint.
+
+##### CSRF for API POST requests
+We know we can disable CSRF checking on the server-side for the mobile POST routes, but that seems unwide.
+We will research best practices for this that do not include turning off the CSRF checking.
+
+
+### Privacy Concerns
+TBD
+
+### Open Questions and Risks
+TBD
+
+### Work Estimates
+N/A
+
+### Alternatives
+#### Alternative to building in vets-api
+##### Build Mobile API entirely outside of vets-api
+Benefits to our team would include autonomy and speed of development, but have some major drawbacks:
+- Lack of reuse - back-end integrations will need to be re-established
+- High LOE when it comes to deployment
+- The work could not be be easily reused by VA.gov
+
+##### Build Mobile API as a proxy in front of vets-api
+Benefits to our team include a degree of autonomy, improved speed of development, but have some definite drawbacks:
+- High LOE when it comes to deployment, although the future VSP container-managed deployment could mitigate this
+- The work could not be be easily reused by VA.gov
+
+#### Path and Protocol Alternatives
+Once we decided on building within vets-api we analyzed our options for how to do so. The short term solutions were to use 
+existing endpoints, version endpoints and make changes there, or create a mobile base path (module) so the mobile routes and 
+auth are isolated. Below is a table listing the sizes for; Lift, or the estimated effort to implement a solution. Latency, 
+the number of calls and size of the payload per mobile screen. Flexibility, how easy it is to make near and long term changes. 
+Ubiquity, the penetration/adoption rate of a technology. And Disruption, how new a technology is to VA and likelihood of 
+requiring a new ATO.
+
+<img src="images/mobile-api/api-decision-matrix.png" alt="API Decision Matrix" width="650" />
```

Comment:
(lol, this table is tough to comment on)

I take mild issue with the "Existing Endpoints" x "Latency" cell --- partly because I think you have overlapping concerns with similar web UIs. if Profile UI takes three calls to get all the data you need... well, that means the web UI takes three calls too and would probably _also_ benefit from optimizing those.

but I agree with the other concerns, you're sending more data than necessary (and by a bit), and being able to more seriously consider options like protobuf optimization in the future could be a big win

Reply:
I'll convert this to a list of alternatives and why we didn't choose them. GraphQL and gRPC can go in future work as they weren't really options for MVP. Agree that the profile optimization should be API wide. Mobile still might have a different serializer if it didn't need all the fields.